designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,440 @@
+
+#. Search for a supported dependency manifest file. If a known manifest is not present, we will abort the build.
+#. Setup: Create build folder
+#. Resolve: Install dependencies
+#. Compile: Optionally, compile the code if necessary
+#. Copy Source: Optionally, Copy Lambda function code to the build folder
+
+Setup step is shared among all runtimes. Other steps in the workflow are implemented differently for each runtime.
+
+Javascript using NPM
+^^^^^^^^^^^^^^^^^^^^
+
+Install dependencies specified by ``package.json`` and copy source files
+
+**Manifest Name**: ``package.json``
+
++-------------+--------------------------------------+
+| Action      | Command                              |
++=============+======================================+
+| Resolve     | ``npm install``                      |
++-------------+--------------------------------------+
+| Compile     | No Op                                |
++-------------+--------------------------------------+
+| Copy Source | Copy files and exclude node_modules  |
++-------------+--------------------------------------+
+
+
+Java using Maven
+^^^^^^^^^^^^^^^^
+
+Let Maven take care of everything
+
+**Manifest Name**: ``pom.xml``
+
++-------------+-----------------+
+| Action      | Command         |
++=============+=================+
+| Resolve     | No Op           |
++-------------+-----------------+
+| Compile     | ``mvn package`` |
++-------------+-----------------+
+| Copy Source | No Op           |
++-------------+-----------------+
+
+Golang using Go CLI
+^^^^^^^^^^^^^^^^^^^
+
+Go's CLI will build the binary.
+
+**Manifest Name**: ``Gopkg.toml``
+
++-------------+---------------------------------------------------+
+| Action      | Command                                           |
++=============+===================================================+
+| Resolve     | ``dep ensure -v``                                 |
++-------------+---------------------------------------------------+
+| Compile     | ``GOOS=linux go build  -ldflags="-s -w" main.go`` |
++-------------+---------------------------------------------------+
+| Copy Source | No Op                                             |
++-------------+---------------------------------------------------+
+
+
+Dotnet using Dotnet CLI
+^^^^^^^^^^^^^^^^^^^^^^^
+
+**Manifest Name**: ``*.csproj``
+
++-------------+----------------------------------------------------------------------------------------+
+| Action      | Command                                                                                |
++=============+========================================================================================+
+| Resolve     | No Op                                                                                  |
++-------------+----------------------------------------------------------------------------------------+
+| Compile     | ``dotnet lambda package --configuration release --output-package $BUILD/package.zip``  |
```

Comment:
cc @sliedig, @blomskog and @pcolazurdo to share their DotNet experience. 

I remember having a chat with @sliedig and @pcolazurdo once that `dotnet lambda package`  includes additional files that are not necessarily needed for Lambda to function -- Perhaps there's a flag or something well known in that space that we can include

Reply:
feedback here would be super helpful!

Reply:
adding Mark @Ubiguchi as he also has great experience in this area for comments

Reply:
The latest version of the Amazon.Lambda.Tools creates a very slim version of the package so I would trust this. The tool creates a zip with all the required dependencies. If project has static files and they are included in the csproj definition, they will be automatically included in the zip

Reply:
I would use the standard microsoft/dotnet docker image to compile. It would need to add dependencies to work with Lambda so a Dockerfile can be included that would add these dependencies on package creation. Not sure what is the model with other languages, but I think this would be a good solution.

Reply:
@pcolazurdo Thanks for the review. Its great to know we are in the right direction here. 

Can you elaborate on your comment about using standard `microsoft/dotnet` image? With native dependencies (like C libraries), we need to compile against a close version of Amazon Linux that Lambda runs. Doesn't dotnet have a similar requirement?

Also, for dependencies, wouldn't you be adding them to your .csproj file or something? Why do you need the Dockerfile?